cpp/include/cudf/table/experimental/row_operators.cuh

```diff
@@ -420,6 +423,279 @@ class self_comparator {
+                             not cudf::is_nested<Element>()),
+              typename... Args>
+    __device__ bool operator()(Args...)
+    {
+      CUDF_UNREACHABLE("Attempted to compare elements of uncomparable types.");
+    }
+
+    template <typename Element, CUDF_ENABLE_IF(cudf::is_nested<Element>())>
+    __device__ bool operator()(size_type const lhs_element_index,
+                               size_type const rhs_element_index) const noexcept
+    {
+      column_device_view lcol = lhs.slice(lhs_element_index, 1);
+      column_device_view rcol = rhs.slice(rhs_element_index, 1);
+      while (is_nested(lcol.type())) {
+        if (nulls) {
+          auto lvalid = make_validity_iterator_safe(lcol);
+          auto rvalid = make_validity_iterator_safe(rcol);
+          if (nulls_are_equal == null_equality::UNEQUAL) {
+            if (thrust::any_of(
+                  thrust::seq, lvalid, lvalid + lcol.size(), thrust::logical_not<bool>()) or
+                thrust::any_of(
+                  thrust::seq, rvalid, rvalid + rcol.size(), thrust::logical_not<bool>())) {
+              return false;
+            }
+          } else {
+            if (not thrust::equal(thrust::seq, lvalid, lvalid + lcol.size(), rvalid)) {
+              return false;
+            }
+          }
+        }
+        if (lcol.type().id() == type_id::STRUCT) {
+          lcol = detail::structs_column_device_view(lcol).sliced_child(0);
+          rcol = detail::structs_column_device_view(rcol).sliced_child(0);
+        } else if (lcol.type().id() == type_id::LIST) {
+          auto l_list_col = detail::lists_column_device_view(lcol);
+          auto r_list_col = detail::lists_column_device_view(rcol);
+
+          auto lsizes = make_list_size_iterator(l_list_col);
+          auto rsizes = make_list_size_iterator(r_list_col);
+          if (not thrust::equal(thrust::seq, lsizes, lsizes + lcol.size(), rsizes)) {
+            return false;
+          }
+
+          lcol = l_list_col.sliced_child();
+          rcol = r_list_col.sliced_child();
+          if (lcol.size() != rcol.size()) { return false; }
+        }
+      }
+
+      auto comp = element_range_comparator{element_comparator{nulls, lcol, rcol, nulls_are_equal}};
+      return type_dispatcher<dispatch_void_if_nested>(lcol.type(), comp, 0, 0, lcol.size());
+    }
+
+   private:
+    struct element_range_comparator {
+      element_comparator const comp;
+
+      template <typename Element, CUDF_ENABLE_IF(cudf::is_equality_comparable<Element, Element>())>
+      __device__ bool operator()(size_type const lhs_element_index,
+                                 size_type const rhs_element_index,
+                                 size_type const size) const noexcept
+      {
+        for (size_type i = 0; i < size; ++i) {
+          bool equal =
+            comp.template operator()<Element>(lhs_element_index + i, rhs_element_index + i);
+          if (not equal) { return false; }
+        }
+        return true;
```

Comment:
```suggestion
   return thrust::equal(thrust::seq, 
                        thrust::counting_iterator{lhs_element_index},                            
                        thrust::counting_iterator{lhs_element_index} + size, 
                        thrust::counting_iterator{rhs_element_index},
                        comp);
```

Reply:
We need to call comp's template operator. The real replacement would be
```c++
        auto comparator = [=] (size_type const lhs_element_index,
                               size_type const rhs_element_index) {
          return comp.template operator()<Element>(lhs_element_index, rhs_element_index);
        };
        thrust::equal(thrust::seq,
                      thrust::make_counting_iterator(lhs_start_index),
                      thrust::make_counting_iterator(lhs_start_index) + size,
                      thrust::make_counting_iterator(rhs_start_index),
                      comparator);

```